File: macos/Tests/ElevateCoreTests/AppVersionTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct AppVersionTests {
    @Test func parsesPlain() {
        let v = AppVersion("1.2.3")
        #expect(v?.major == 1 && v?.minor == 2 && v?.patch == 3)
    }

    @Test func parsesLeadingV() {
        let v = AppVersion("v1.2.3")
        #expect(v?.major == 1 && v?.minor == 2 && v?.patch == 3)
    }

    @Test func parsesBuildMetadata() {
        let v = AppVersion("1.2.3+45")
        #expect(v?.major == 1 && v?.minor == 2 && v?.patch == 3)
    }

    @Test func parsesMissingPatchAsZero() {
        let v = AppVersion("1.2")
        #expect(v?.major == 1 && v?.minor == 2 && v?.patch == 0)
    }

    @Test func rejectsGarbage() {
        #expect(AppVersion("") == nil)
        #expect(AppVersion("abc") == nil)
        #expect(AppVersion("1.x") == nil)
    }

    @Test func comparable() {
        #expect(AppVersion("1.2.3")! < AppVersion("1.2.4")!)
        #expect(AppVersion("1.2.3")! < AppVersion("1.3.0")!)
        #expect(AppVersion("1.2.3")! < AppVersion("2.0.0")!)
        #expect(AppVersion("1.2.3")! == AppVersion("1.2.3")!)
        #expect(!(AppVersion("1.2.3")! < AppVersion("1.2.3")!))
    }

    @Test func isNewerTrueWhenLatestGreater() {
        #expect(AppVersion.isNewer(latestTag: "v1.1.0", current: "1.0.0"))
        #expect(AppVersion.isNewer(latestTag: "1.0.1", current: "1.0.0"))
    }

    @Test func isNewerFalseWhenEqual() {
        #expect(!AppVersion.isNewer(latestTag: "v1.0.0", current: "1.0.0"))
    }

    @Test func isNewerFalseWhenOlder() {
        #expect(!AppVersion.isNewer(latestTag: "v0.9.0", current: "1.0.0"))
    }

    @Test func isNewerFalseWhenEitherFailsToParse() {
        #expect(!AppVersion.isNewer(latestTag: "garbage", current: "1.0.0"))
        #expect(!AppVersion.isNewer(latestTag: "v1.0.0", current: "garbage"))
        #expect(!AppVersion.isNewer(latestTag: "garbage", current: "also-garbage"))
    }
}

File: windows/tests/Elevate.Core.Tests/AppVersionTests.cs
using Elevate.Core.Support;
using FluentAssertions;

namespace Elevate.Core.Tests;

/// <summary>Port of the Swift <c>AppVersionTests</c>.</summary>
public class AppVersionTests
{
    [Theory]
    [InlineData("1.2.3", 1, 2, 3)]
    [InlineData("v1.2.3", 1, 2, 3)]
    [InlineData("1.2.3+45", 1, 2, 3)]
    [InlineData("1.2", 1, 2, 0)]
    public void Parses(string text, int major, int minor, int patch) =>
        AppVersion.Parse(text).Should().Be(new AppVersion(major, minor, patch));

    [Theory]
    [InlineData("")]
    [InlineData("abc")]
    [InlineData("1.x")]
    [InlineData("1")]
    [InlineData("1.2.3.4")]
    public void RejectsGarbage(string text) => AppVersion.Parse(text).Should().BeNull();

    [Fact]
    public void Comparable()
    {
        (AppVersion.Parse("1.2.3")!.Value < AppVersion.Parse("1.2.4")!.Value).Should().BeTrue();
        (AppVersion.Parse("1.2.3")!.Value < AppVersion.Parse("1.3.0")!.Value).Should().BeTrue();
        (AppVersion.Parse("1.2.3")!.Value < AppVersion.Parse("2.0.0")!.Value).Should().BeTrue();
        AppVersion.Parse("1.2.3").Should().Be(AppVersion.Parse("1.2.3"));
        (AppVersion.Parse("1.2.3")!.Value < AppVersion.Parse("1.2.3")!.Value).Should().BeFalse();
    }

    [Fact]
    public void IsNewerTrueWhenLatestGreater()
    {
        AppVersion.IsNewer("v1.1.0", "1.0.0").Should().BeTrue();
        AppVersion.IsNewer("1.0.1", "1.0.0").Should().BeTrue();
    }

    [Fact]
    public void IsNewerFalseWhenEqual() => AppVersion.IsNewer("v1.0.0", "1.0.0").Should().BeFalse();

    [Fact]
    public void IsNewerFalseWhenOlder() => AppVersion.IsNewer("v0.9.0", "1.0.0").Should().BeFalse();

    [Fact]
    public void IsNewerFalseWhenEitherFailsToParse()
    {
        AppVersion.IsNewer("garbage", "1.0.0").Should().BeFalse();
        AppVersion.IsNewer("v1.0.0", "garbage").Should().BeFalse();
        AppVersion.IsNewer("garbage", "also-garbage").Should().BeFalse();
    }

    [Fact]
    public void WindowsTagPrefixIsNotAVersion() => AppVersion.Parse("windows-v1.2.3").Should().BeNull();
}
